### PR TITLE
Implement NFT Class reveal queue abstract

### PR DESCRIPTION
**Pull request by hochiw**
## Features
- Scaffolded ClassRevealQueue map via ignite cli
- Created methods to interact with the queue 
- Created an abstract abci endblocker to process the queue at the end of the block, for use after oursky/likecoin-chain#206 
- Enqueue class for reveal on blindbox creation
- Update or enqueue class for reveal on blindbox update

## Testcases
- Created testcase for inserting entries into the queue
- Created testcase for deleting entry from the queue
- Created testcase for updating an entry from the queue 
- Created testcase to check the sorting of the queue
- Created testcase scaffold for the abci, for use after oursky/likecoin-chain#206 

refs oursky/likecoin-chain#205

### CODE REVIEW

File: proto/likenft/class_reveal_queue.proto [hunk 1]
```diff
@@ -3,9 +3,11 @@ package likecoin.likechain.likenft;
 
 option go_package = "github.com/likecoin/likechain/x/likenft/types";
 
+import "gogoproto/gogo.proto";
+import "google/protobuf/timestamp.proto";
+
 message ClassRevealQueue {
-  string revealTime = 1; 
+  google.protobuf.Timestamp revealTime = 1 [(gogoproto.stdtime) = true];
```
Comment (chihimng) [inline proto/likenft/class_reveal_queue.proto:10]: set nullable to false as well? so the generated go type will not be pointer

File: proto/likenft/event.proto [hunk 2]
```diff
@@ -15,6 +15,10 @@ message EventUpdateClass {
   string parentAccount = 3;
 }
 
+message EventRevealClass {
+  string classId = 1; 
+}
```
Comment (chihimng) [inline proto/likenft/event.proto:20]: can we add class parent fields here for easier listening?

File: x/likenft/abci.go [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+	keeper.IterateClassRevealQueue(ctx, func(classId string) bool {
+		// Note: Create public functions from keeper files and access them here instead of mutating directly
+
+		// TODO: Mint unclaimed tokens supply to class owner #206
+
+		// TODO: Reveal nft class #206
+
+		// TODO: Remove class from queue #206
+
+		ctx.EventManager().EmitTypedEvent(&types.EventRevealClass{
```
Comment (chihimng) [inline x/likenft/abci.go:28]: this should be part of business logic, not queue level, but let's leave it here for now

Comment (chihimng) [reply x/likenft/abci.go:28]: add a todo for refactoring maybe?

File: x/likenft/abci.go [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+			ClassId: classId,
+		})
+
```
Comment (chihimng) [inline x/likenft/abci.go:31]: btw, can we see how to handle exceptions? like where to log & make sure errors will only cause the entry to be skipped and not revert the whole batch?

on exceptions, how to let user trigger retry?

**Review by elise-ng — CHANGES_REQUESTED** [review]

File: x/likenft/abci.go (hunk 4 above)
Comment (hochiw) [reply x/likenft/abci.go:31]: errors should be logged by the logger and the entry shall be retried when the next block arrives if it's not removed from the queue

Comment (chihimng) [reply x/likenft/abci.go:32]: - let's add success status & error message at the reveal event
- we should dequeue if failed, and consider if we need to expose a manual reveal method (depending on the failure modes, after we implement the logic)
- i suppose the logic will return a error, but can we capture all unhandled panic / exceptions here?

**Comment by chihimng** [comment]
Ref issue #205

**Comment by chihimng** [comment]
@hochiw you missed `Update queued entry on change` ? on the issue checklist

**Comment by chihimng** [comment]
> 

    let's add success status & error message at the reveal event
    we should dequeue if failed, and consider if we need to expose a manual reveal method (depending on the failure modes, after we implement the logic)
    i suppose the logic will return a error, but can we capture all unhandled panic / exceptions here?


I have implemented these just now

**Comment by hochiw** [comment]
> @hochiw you missed `Update queued entry on change` ? on the issue checklist

mb you are right, added commits for this

**Comment by chihimng** [comment]
@hochiw

Refactored the protos and some logic, also added check for mintable count, can you help cross check?

File: x/likenft/keeper/msg_server_update_class.go [hunk 5]
```diff
@@ -82,18 +82,17 @@ func (k msgServer) UpdateClass(goCtx context.Context, msg *types.MsgUpdateClass)
...
+	if originalConfig.IsBlindBox() {
+		k.RemoveClassRevealQueueEntry(ctx, originalConfig.BlindBoxConfig.RevealTime, class.Id)
+	}
+
+	// Enqueue new reveal schedule
+	if updatedConfig.IsBlindBox() {
 		k.SetClassRevealQueueEntry(ctx, types.ClassRevealQueueEntry{
 			ClassId:    class.Id,
 			RevealTime: updatedConfig.BlindBoxConfig.RevealTime,
 		})
```
Comment (hochiw) [inline x/likenft/keeper/msg_server_update_class.go:94]: so the update function is not used anymore? but i guess we can keep it just in case

**Comment by hochiw** [comment]
@chihimng seems okay

**Review by elise-ng — APPROVED** [review]

**Comment by chihimng** [comment]
ran test locally, passes; handle ci timeout later

**Pull request merged**